[AArch64] merge index address with large offset into base address

vfdff opened the pull request:
A case for this transformation, https://gcc.godbolt.org/z/nhYcWq1WE
```
Fold
  mov     w8, #56952
  movk    w8, #15, lsl #16
  ldrb    w0, [x0, x8]
into
  add     x0, x0, 1036288
  ldrb    w0, [x0, 3704]
```
Only support single use base, multi-use scenes are supported by PR74046.
Fix https://github.com/llvm/llvm-project/issues/71917

efriedma-quic commented:
This seems like it would be simpler to do at the point where we're computing the addressing mode in the first place (in ISelDAGToDAG).

vfdff commented:
> This seems like it would be simpler to do at the point where we're computing the addressing mode in the first place (in ISelDAGToDAG).

Thanks for your idea @efriedma-quic
I tried to refactor it in `AArch64DAGToDAGISel::SelectAddrModeXRO` according your idea, and find  I can only modify the **Base** and **Offset** of the load/store instruction.
Therefore, the following format is generated after the adjustment. The instruction LDRWroX here requires [reg + reg], so an error is reported, Is there a way to adjust the opcode?
```
%1:gpr32 = MOVi32imm 4160992
%2:gpr64 = SUBREG_TO_REG 0, killed %1:gpr32, %subreg.sub_32
%3:gpr32 = LDRWroX %0:gpr64common, %2:gpr64, 0, 0
 ---> 
%1:gpr64sp = ADDXri %0:gpr64common, 1012, 12
%2:gpr32 = LDRWroX killed %1:gpr64sp, 15840, 0, 0  // The 2nd operand 15840 required a reg
```

vfdff committed:
[AArch64] merge index address with large offset into base address

A case for this transformation, https://gcc.godbolt.org/z/nhYcWq1WE
```
Fold
  mov     w8, #56952
  movk    w8, #15, lsl #16
  ldrb    w0, [x0, x8]
into
  add     x0, x0, 1036288
  ldrb    w0, [x0, 3704]
```
Only support single use base, multi-use scenes are supported by PR74046.
Fix https://github.com/llvm/llvm-project/issues/71917

TODO: support the multiple-uses with reuseing common base offset.
https://gcc.godbolt.org/z/Mr7srTjnz

vfdff commented:
hi @efriedma-quic, I refactor that in `AArch64DAGToDAGISel::SelectAddrModeIndexed` to fix the above crash.

davemgreen commented on llvm/test/CodeGen/AArch64/arm64-addrmode.ll:
```diff
@@ -209,3 +209,28 @@ define void @t17(i64 %a) {
+; CHECK-NEXT:    add x8, x0, #507, lsl #12 // =2076672
+; CHECK-NEXT:    ldrsh w0, [x8, #3312]
+; CHECK-NEXT:    ret
+  %arrayidx = getelementptr inbounds i16, ptr %a, i64 1039992
```
What happens if this is a i8 scale and the immediate is odd?

It's probably worth adding a few more tests with different type, having multiple offsets in a single test, etc. Thanks

vfdff replied:
Prevent transforming when the immediate is odd, add a related test `LdOffset_i16_odd_offset` for that.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -1110,6 +1142,24 @@ bool AArch64DAGToDAGISel::SelectAddrModeIndexed(SDValue N, unsigned Size,
+      //     ADD  BaseReg, WideImmediate & 0x0fff000
+      //     LDR  X2, [BaseReg, WideImmediate & 0x0fff]
+      SDValue LHS = N.getOperand(0);
+      if (isPreferredBaseAddrMode(RHSC)) {
```
I think there needs to be a check for Scale here too.

vfdff replied:
Thanks, I'll add the `check for Scale`, but this merge doesn't change odd or even.

davemgreen reviewed (commented):
It might be worth doing this as a (late) AArch64ISelLowering DAG combine so that the add gets shared between multiple load/store instructions.

vfdff commented:
> It might be worth doing this as a (late) AArch64ISelLowering DAG combine so that the add gets shared between multiple load/store instructions.
* Yes, the current solution does not implement the reuse of add, eg:
```
define i64 @LdOffset_i64_multi_offset(ptr %a) {
; CHECK-LABEL: LdOffset_i64_multi_offset:
; CHECK:       // %bb.0: // %entry
; CHECK-NEXT:    add x8, x0, #2031, lsl #12 // =8318976
; CHECK-NEXT:    add x9, x0, #2031, lsl #12 // =8318976
; CHECK-NEXT:    add x8, x8, #960
; CHECK-NEXT:    ldr x9, [x9, #960]
; CHECK-NEXT:    ldr x8, [x8, #2056]
; CHECK-NEXT:    add x0, x8, x9
; CHECK-NEXT:    ret
entry:
  %arrayidx = getelementptr inbounds i64, ptr %a, i64 1039992
  %0 = load i64, ptr %arrayidx, align 8
  %arrayidx1 = getelementptr inbounds i64, ptr %a, i64 1040249
  %1 = load i64, ptr %arrayidx1, align 8
  %add = add nsw i64 %1, %0
  ret i64 %add
}
```
* In that case, I think it's reasonable to go back to my previous plan? （I'll add more test cases if it's right direction).
   I don't prefer to `DAG combine` because it seems we need similar code to address `load` and `store` separately
  https://github.com/llvm/llvm-project/commit/bff3b81d921a7c69fd8a50dcbf7cd44776c73d19

vfdff commented:
> * performLOADCombine

davemgreen commented:
> I don't prefer to DAG combine because it seems we need similar code to address load and store separately, may be performLOADCombine and performStoreCombine.

Could it share a similar method called from both performLOADCombine and performStoreCombine? I'm not sure if the opposite transform exists in DAG, but if it doesn't it should help only produce a single add.

vfdff commented:
hi @davemgreen 
  * I checked the `DAG combine` is placed before the `AArch64DAGToDAGISel`, and it doesn't hold the `machine IR`）, so it only uses a register `t4` to specify the destination address.
```
t4: i64 = add nuw t2, Constant:i64<1039992>
    t12: i32,ch = load<(load (s8) from %ir.arrayidx, !tbaa !6), zext from i8> t0, t4, undef:i64
```
* When I split the const offset into 2 immediate numbers, then it will merge back in `DAGCombiner::visitADD`.
```
t15: i64 = add t2, Constant:i64<1036288>
      t17: i64 = add t15, Constant:i64<3704>
    t18: i32,ch = load<(load (s8) from %ir.arrayidx, !tbaa !6), zext from i8> t0, t17, undef:i64
```
* then I try to stop above the back merging in `DAGCombiner::visitADD->..->FoldConstantArithmetic`, it still can't reuse the
  add instruction,  i.e: it has **same** output to the method done in  `AArch64DAGToDAGISel`.
  test with multiple offsets in one case : https://gcc.godbolt.org/z/dE8dETcvK
  ```
	add	x8, x0, #1015, lsl #12          // =4157440
	add	x9, x0, #1015, lsl #12          // =4157440
	add	x8, x8, #2528
	ldr	w9, [x9, #2528]
	ldr	w8, [x8, #8]
	add	w0, w8, w9
  ```

 so need adaptation support in **CodeGenPrepare::optimizeMemoryInst** ?

davemgreen commented:
Ah I see. Like this: https://gcc.godbolt.org/z/4Ph36xdEv? It sounds like that might work, so long as the operands have multiple uses it shouldn't just transform it back to a single add. It might need something like the current patch to handle single-use cases, with something different for multiple-uses.

vfdff commented:
Thanks @davemgreen for your idea.
Based on the above discussion, I restrict the single use for the 1st improvement,

vfdff commented:
Add the 2nd patch to fix remainder multi-use common base

vfdff commented:
update to fix Transforms/CodeGenPrepare/AArch64/large-offset-gep.ll

davemgreen commented on llvm/test/CodeGen/AArch64/arm64-addrmode.ll:
```diff
@@ -209,3 +209,88 @@ define void @t17(i64 %a) {
   ret void
 }
+
+; https://gcc.godbolt.org/z/ErhhdxMv3
```
It's probably best to remove the godbolt links from the tests.

vfdff replied:
Done

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -15977,6 +15977,18 @@ bool AArch64TargetLowering::isLegalAddressingMode(const DataLayout &DL,
+// 12bits are same.
+int64_t AArch64TargetLowering::getPreferBaseOffset(int64_t MinOffset,
+                                                   int64_t MaxOffset) const {
+  if (MinOffset >> 12 == MaxOffset >> 12 && MinOffset >> 24 == 0) {
```
I think this is assuming a uimm12s1 addressing mode for the load/store? Should it be more precise about the type being used, and maybe make use of isLegalAddressingMode if it can?

vfdff replied:
thanks, I updated with API `isLegalAddImmediate`

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -1069,6 +1069,42 @@ bool AArch64DAGToDAGISel::SelectAddrModeIndexedBitWidth(SDValue N, bool IsSigned
 }
 
+// 16-bit optionally shifted immediates are legal for single mov.
+static bool isLegalSingleMOVImmediate(int64_t Immed) {
```
Can this make use of expandMOVImm?
```
    AArch64_IMM::expandMOVImm(UImm, BitSize, Insn);
    if (Insn.size() != 1)
      return;
```

vfdff replied:
Apply your comment, thanks

davemgreen reviewed (commented):
It is probably best to split the CGP and DAG parts out into separate patches if we can. I'm seeing some failures in the test-suite, although the look like they might be post-isel so there might be something else going on. I'm still not 100% sure how to structure this without causing regressions in places.

vfdff commented:
> It is probably best to split the CGP and DAG parts out into separate patches if we can. I'm seeing some failures in the test-suite, although the look like they might be post-isel so there might be something else going on. I'm still not 100% sure how to structure this without causing regressions in places.

hi @davemgreen, I spilt the 2nd patch into [pr74046](https://github.com/llvm/llvm-project/pull/74046) according your idea, and that I'll rebase this patch after that patch is merged because the `OneUse` checking can be avoid for the 1st patch if the 2nd is merged first.

davemgreen commented:
I still seem to see some performance issues from this, in the TSVC benchmarks in the llvm test suite. Do you see the same thing?

davemgreen commented:
It looks like it was running with `-O3 -flto -mcpu=neoverse-v1`, although I can reproduce without the -mcpu.
`bin/clang -target aarch64 -O3 llvm-test-suite/MultiSource/Benchmarks/TSVC/StatementReordering-flt/tsc.c`

I think it is in these loops that look like:
```
.LBB6_3:                               |  10079 .LBB6_3:                          
        ldr     s1, [x9, x23]          |  10082         add     x10, x9, #93, lsl #12   
        ldr     s2, [x9, x24]          |  10083         add     x11, x9, #62, lsl #12   
        subs    x8, x8, #1             |  10084         subs    x8, x8, #1              
---------------------------------------|  10085         ldr     s1, [x11, #2100]        
---------------------------------------|  10086         ldr     s2, [x10, #3156]        
        fmadd   s0, s1, s2, s0         |  10087         fmadd   s0, s1, s2, s0          
        str     s0, [x9, #4]!          |  10088         str     s0, [x9, #4]!           
        add     x10, x9, x22           |  10089         add     x10, x9, #125, lsl #12  
        ldr     s0, [x10, #4]          |  10090         add     x11, x9, x22            
        ldr     s1, [x9, x25]          |  10091         ldr     s1, [x10, #128]         
---------------------------------------|  10092         add     x10, x9, #31, lsl #12   
---------------------------------------|  10093         ldr     s0, [x11, #4]           
        fmsub   s0, s1, s2, s0         |  10094         fmsub   s0, s1, s2, s0          
        str     s0, [x10]              |  10095         str     s0, [x10, #1040]        
        b.ne    .LBB6_3                |  10096         b.ne    .LBB6_3                 
```

vfdff commented:
Thanks for your help, I can reproduce the regression now, and a simplified case [s211](https://gcc.godbolt.org/z/z4rYaj8qn) shows that.
a) without this changes, the const `MOVi32imm constImm` is a loop Invariant, so it is hoisted out the loop.
b) With this changes, the `%12:grp64sp = ADDxri %2:gpr64sp 93, 12` is not a loop Invariant, which checked by the  **MachineLoop::isLoopInvariant**, then it is keep in the loop kernel.
   ```
     // If the loop contains the definition of an operand, then the instruction isn't loop invariant.
    if (contains(MRI->getVRegDef(Reg)))
      return false;
  ```

vfdff commented:
hi @efriedma-quic @efriedma-quic 
  Because I don't fix out how the check the loop invariant in `AArch64DAGToDAGISel` to avoid some regression cases, so I roll back solution to the 1st method, rebased the patch in [#75343](https://github.com/llvm/llvm-project/pull/75343)

vfdff commented:
close because the new solution: https://github.com/llvm/llvm-project/pull/75343
